### PR TITLE
feat(checkout): CHECKOUT-7536 Load extensions

### PR DESCRIPTION
**Pull request by bc-peng**
## What?
- Retrieve all extension configurations on a store.

- Integrate them seamlessly into the existing checkout state.

## Why?
Efficiently manage and utilize extensions.

## Testing / Proof
- CI checks.
- Manual testing.
   image

@bigcommerce/checkout @bigcommerce/payments

### CODE REVIEW

**Review by animesh1987 — COMMENTED** [review]
Nice work overall. Had a question, have we thought about moving `extension` package to it's own separate package out of core?

**Comment by bc-peng** [comment]
> Nice work overall. Had a question, have we thought about moving `extension` package to it's own separate package out of core?

Thank you for your suggestion.
I'm not sure I see the benefits, given that the majority of the files are already defined here.
If we decide to make it a package, we'll need to figure out how to provide access to the checkout service and state similar to the payment integration API. It may be tricky but doable.

File: packages/core/src/extension/extension-action-creator.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+export class ExtensionActionCreator {
+    constructor(private _requestSender: ExtensionRequestSender) {}
+
+    loadExtensions(
+        options?: RequestOptions,
+    ): ThunkAction<ExtensionAction, InternalCheckoutSelectors> {
+        return (store) =>
+            Observable.create((observer: Observer<ExtensionAction>) => {
+                const state = store.getState();
+                const cart = state.cart.getCartOrThrow();
```
Comment (animesh1987) [inline packages/core/src/extension/extension-action-creator.ts:19]: Do we need to pass cartId for extensions? Aren't they something that belong at store level?

File: packages/core/src/extension/extension-request-sender.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import { Extension } from './extension';
+
+export const EXTENSIONS_API_URL = '/api/storefront/checkout-extensions';
+
+export class ExtensionRequestSender {
+    constructor(private _requestSender: RequestSender) {}
+
+    loadExtensions({ timeout, params }: RequestOptions = {}): Promise<Response<Extension[]>> {
+        return this._requestSender.get(EXTENSIONS_API_URL, {
+            timeout,
```
Comment (animesh1987) [inline packages/core/src/extension/extension-request-sender.ts:19]: What does this timeout do?

File: packages/core/src/extension/extension-request-sender.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            headers: {
+                Accept: ContentType.JsonV1,
+                'X-API-INTERNAL': INTERNAL_USE_ONLY,
```
Comment (animesh1987) [inline packages/core/src/extension/extension-request-sender.ts:22]: I remember there was discussion offline, do we need to add this?

File: packages/core/src/extension/extension-request-sender.ts (hunk 2 above)
Comment (bc-peng) [reply packages/core/src/extension/extension-request-sender.ts:14]: I think we offer `timeout`for almost all requests, it is a standard option in `@bigcommerce/request-sender`.
> Provide this option if you want to cancel or time out the request. If the timeout object completes before the request, the request will be cancelled.

File: packages/core/src/extension/extension-request-sender.ts (hunk 3 above)
Comment (bc-peng) [reply packages/core/src/extension/extension-request-sender.ts]: Removed.

File: packages/core/src/extension/extension-action-creator.ts (hunk 1 above)
Comment (bc-peng) [reply packages/core/src/extension/extension-action-creator.ts]: I put it here to match with loading cart and payment methods.
Removed.
On another note, when we render an extension, we'll pass `cartId` and `extensionId` to an iframe as params.

File: packages/core/src/extension/extension-request-sender.ts (hunk 2 above)
Comment (animesh1987) [reply packages/core/src/extension/extension-request-sender.ts:14]: Think seems like a copy paste issue, as the client will need to pass `timeout` as options, which we rarely do 🤔

File: packages/core/src/extension/extension-action-creator.ts (hunk 1 above)
Comment (animesh1987) [reply packages/core/src/extension/extension-action-creator.ts]: Yes that can we done at a different layer, if we don't need information as part of action creation we should avoid it.

**Review by animesh1987 — APPROVED** [review]
Nice work 👍

**Pull request merged**